### PR TITLE
[multi-dut] add infrastructure to update dut port to ptf port index map

### PR DESCRIPTION
**Pull request by yxieca**
### Description of PR

Summary:
Fixes # (issue)

### Type of change

- [ ] Bug fix
- [x] Testbed and Framework(new/improvement)
- [x] Test case(new/improvement)

#### What is the motivation for this PR?
In the single DUT scenario, the DUT interface indeces exactly match with the ptf port indeces. However, in multi-dut scneario, this is not always true. Multiple DUTs are sharing the same ptf host, therefore their interface indeces need to be translated to ptf port indeces according to the topology definition.

#### How did you do it?
In this change, a translate map is captured from topology definition. And member method get_extended_minigraph_facts() is added to SonicHost to help getting the minigrap facts and then extend it with ptf port indices with mg_facts['minigraph_ptf_indeces'].

Changed test_lag_2 to use the new method to map to the correct ptf port.

Signed-off-by: Ying Xie <ying.xie@microsoft.com>

#### How did you verify/test it?
Tested the infrastructure with updated lag_2 test.

platform linux2 -- Python 2.7.12, pytest-4.6.5, py-1.9.0, pluggy-0.13.1
ansible: 2.8.12
rootdir: /var/src/sonic-mgmt/tests, inifile: pytest.ini
plugins: ansible-2.2.2, forked-1.3.0, xdist-1.28.0, html-1.22.1, repeat-0.8.0, metadata-1.10.0
collected 24 items                                                                                                                                                                       

pc/test_lag_2.py::test_lag[str2-7050cx3-acs-06|PortChannel0001-single_lag] PASSED                                                                                                  [  4%]
pc/test_lag_2.py::test_lag[str2-7050cx3-acs-06|PortChannel0001-lacp_rate] PASSED                                                                                                   [  8%]
pc/test_lag_2.py::test_lag[str2-7050cx3-acs-06|PortChannel0001-fallback] SKIPPED                                                                                                   [ 12%]
pc/test_lag_2.py::test_lag[str2-7050cx3-acs-06|PortChannel0003-single_lag] PASSED                                                                                                  [ 16%]
pc/test_lag_2.py::test_lag[str2-7050cx3-acs-06|PortChannel0003-lacp_rate] PASSED                                                                                                   [ 20%]
pc/test_lag_2.py::test_lag[str2-7050cx3-acs-06|PortChannel0003-fallback] SKIPPED                                                                                                   [ 25%]
pc/test_lag_2.py::test_lag[str2-7050cx3-acs-06|PortChannel0002-single_lag] PASSED                                                                                                  [ 29%]
pc/test_lag_2.py::test_lag[str2-7050cx3-acs-06|PortChannel0002-lacp_rate] PASSED                                                                                                   [ 33%]
pc/test_lag_2.py::test_lag[str2-7050cx3-acs-06|PortChannel0002-fallback] SKIPPED                                                                                                   [ 37%]
pc/test_lag_2.py::test_lag[str2-7050cx3-acs-06|PortChannel0004-single_lag] PASSED                                                                                                  [ 41%]
pc/test_lag_2.py::test_lag[str2-7050cx3-acs-06|PortChannel0004-lacp_rate] PASSED                                                                                                   [ 45%]
pc/test_lag_2.py::test_lag[str2-7050cx3-acs-06|PortChannel0004-fallback] SKIPPED                                                                                                   [ 50%]
pc/test_lag_2.py::test_lag[str2-7050cx3-acs-07|PortChannel0001-single_lag] PASSED                                                                                                  [ 54%]
pc/test_lag_2.py::test_lag[str2-7050cx3-acs-07|PortChannel0001-lacp_rate] PASSED                                                                                                   [ 58%]
pc/test_lag_2.py::test_lag[str2-7050cx3-acs-07|PortChannel0001-fallback] SKIPPED                                                                                                   [ 62%]
pc/test_lag_2.py::test_lag[str2-7050cx3-acs-07|PortChannel0003-single_lag] PASSED                                                                                                  [ 66%]
pc/test_lag_2.py::test_lag[str2-7050cx3-acs-07|PortChannel0003-lacp_rate] PASSED                                                                                                   [ 70%]
pc/test_lag_2.py::test_lag[str2-7050cx3-acs-07|PortChannel0003-fallback] SKIPPED                                                                                                   [ 75%]
pc/test_lag_2.py::test_lag[str2-7050cx3-acs-07|PortChannel0002-single_lag] PASSED                                                                                                  [ 79%]
pc/test_lag_2.py::test_lag[str2-7050cx3-acs-07|PortChannel0002-lacp_rate] PASSED                                                                                                   [ 83%]
pc/test_lag_2.py::test_lag[str2-7050cx3-acs-07|PortChannel0002-fallback] SKIPPED                                                                                                   [ 87%]
pc/test_lag_2.py::test_lag[str2-7050cx3-acs-07|PortChannel0004-single_lag] PASSED                                                                                                  [ 91%]
pc/test_lag_2.py::test_lag[str2-7050cx3-acs-07|PortChannel0004-lacp_rate] PASSED                                                                                                   [ 95%]
pc/test_lag_2.py::test_lag[str2-7050cx3-acs-07|PortChannel0004-fallback] SKIPPED                                                                                                   [100%]

------------------------------------------------------------------------ generated xml file: /tmp/logs-01/tr.xml -------------------------------------------------------------------------

### CODE REVIEW

File: tests/conftest.py [hunk 1]
```diff
@@ -101,6 +102,64 @@ def get_testbed_type(self, topo_name):
...
+
+        """
+        m = re.match("(\d+)\.(\d+)@(\d+)", port)
+        (dut_index, port_index, ptf_index) = (int(m.group(1)), int(m.group(2)), int(m.group(3)))
+
+        return (dut_index, port_index, ptf_index)
+
+    def calculate_ptf_index_map(self, line):
+        map = defaultdict()
+        if len(line['duts']) <= 1:
```
Comment (lguohan) [inline tests/conftest.py:119]: we may want to move to the multi-duts format in topology file in the future.

Comment (yxieca) [reply tests/conftest.py:119]: I think we can move generic information to topology file. But here line['duts'] are defining the specific information of the testbed, it provides list of hostnames for the duts in this testbed. This part probably have to stay in the testbed definition.

Comment (yxieca) [reply tests/conftest.py:119]: Did you mean ignore this field and use topology property 'dut_num' for this test instead?

**Comment by yxieca** [comment]
retest vsimage please

File: tests/conftest.py (hunk 1 above)
Comment (yxieca) [reply tests/conftest.py]: Per our discussion offline. Removed the dut count check.

**Review by lguohan — APPROVED** [review]

**Review by Blueve — APPROVED** [review]

**Pull request merged**